File: algorithms/hard/number_of_visible_people_in_a_queue.ex
defmodule Solution do
  @spec can_see_persons_count(heights :: [integer]) :: [integer]
  def can_see_persons_count(heights) do
    can_see_persons_count(Enum.reverse(heights), [], [])
  end

  def can_see_persons_count([], _st, ans), do: ans

  def can_see_persons_count([h | hs], [], ans) do
    can_see_persons_count(hs, [h], [0 | ans])
  end

  def can_see_persons_count([h | hs], [s | ss] = st, ans) do
    if h < s do
      can_see_persons_count(hs, [h | st], [1 | ans])
    else
      {popped, remained} = Enum.split_while(st, &(&1 < h))
      n = Enum.count(popped)
      additional = (remained == [] && 0) || 1
      can_see_persons_count(hs, [h | remained], [n + additional | ans])
    end
  end
end

File: algorithms/hard/number_of_visible_people_in_a_queue.rs
impl Solution {
    pub fn can_see_persons_count(heights: Vec<i32>) -> Vec<i32> {
        let mut st = Vec::new();
        let mut ans = Vec::new();

        for h in heights.into_iter().rev() {
            if st.is_empty() {
                st.push(h);
                ans.push(0);
                continue;
            }

            if h < st[st.len() - 1] {
                st.push(h);
                ans.push(1);
                continue;
            }

            let mut n = 0;
            while !st.is_empty() && st[st.len() - 1] < h {
                n += 1;
                st.pop();
            }

            if !st.is_empty() {
                n += 1;
            }

            st.push(h);
            ans.push(n);
        }

        ans.into_iter().rev().collect()
    }
}
